src/include/storage/index/base_hash_index.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include <shared_mutex>
+#include <utility>
+
+#include "common/types/types.h"
+#include "function/hash/hash_functions.h"
+#include "hash_index_header.h"
+#include "hash_index_slot.h"
+#include "transaction/transaction.h"
+
+namespace kuzu {
+namespace storage {
+
+enum class ChainedSlotsAction : uint8_t { LOOKUP_IN_SLOTS, DELETE_IN_SLOTS, FIND_FREE_SLOT };
+
+enum class SlotType : uint8_t { PRIMARY = 0, OVF = 1 };
+
+struct SlotInfo {
+    slot_id_t slotId{UINT64_MAX};
+    SlotType slotType{SlotType::PRIMARY};
+};
+
+static constexpr common::page_idx_t INDEX_HEADER_ARRAY_HEADER_PAGE_IDX = 0;
+static constexpr common::page_idx_t P_SLOTS_HEADER_PAGE_IDX = 1;
+static constexpr common::page_idx_t O_SLOTS_HEADER_PAGE_IDX = 2;
+static constexpr common::page_idx_t NUM_HEADER_PAGES = 3;
+static constexpr uint64_t INDEX_HEADER_IDX_IN_ARRAY = 0;
+
+template<typename T, typename S = T>
+class BaseHashIndex {
+public:
+    explicit BaseHashIndex() : slotCapacity(getSlotCapacity<T>()) {}
```

Comment:
For `getSlotCapacity`, it seems we didn't refresh it for newly supported data types. (I cannot add comment at the place where the code is, so note it here).
```cpp
template<typename T>
static constexpr uint8_t getSlotCapacity() {
    if (std::is_same<T, int64_t>::value) {
        return common::HashIndexConstants::INT64_SLOT_CAPACITY;
    } else {
        return common::HashIndexConstants::STRING_SLOT_CAPACITY;
    }
}
```

Reply:
Right.